+str add in-line inspect operator for sideeffecting

ktoso opened the pull request:
After seeing yet another (few) codebases this week (and wanting it in pet project myself a few times) that use `.map { x => { report(x); x } }` quite a bit I finally decided to PR this.

It's the in-line version of foreach; implemented by map, so no new implementation.

Resolves https://github.com/akka/akka/issues/23512

ktoso committed:
+str add in-line wireTap operator for sideeffecting

johanandren commented:
I'm -1 on this, but you already know this. 

In addition to that I think `foreach` is a bad name because it already exists as `runForeach` and `Sink.foreach` plus that a foreach in code is the "terminating" loop, it doesn't allow for further processing of every element, so if we should have it, I think it should have a different name.

hepin1989 commented:
RxJava have `doOnEach`, I think that would be a better name?

raboof commented:
I found `foreach` confusing because I'd expect it to return `Unit`, something with `do` (or perhaps just plainly `do`?) might indeed be nice?

tarsa commented:
`foreach` is confusing for me as well, i.e. when talking about `foreach` in Akka Streams context it often wouldn't be immediately obvious whether someone is talking about terminal `foreach` or intermediate `foreach`.

Also we should suggest that the function shouldn't be used for mutating the element. Therefore I propose name like `inspect` instead of `foreach`.

ktoso commented:
Interesting argument and proposal! Thanks, will ponder it more, this sounds pretty interesting indeed :)

hepin1989 commented:
It's called `doOnNext` in RxJava, and I would like to call it `doOnPush`.

ktoso commented:
As mentioned, `inspect` expresses the intent well. I amended the PR.

tarsa commented on akka-docs/src/main/paradox/stream/stages-overview.md:
```diff
@@ -719,6 +719,22 @@ Transform each element in the stream by calling a mapping function with it and p
 
 ---------------------------------------------------------------
 
+### foreach 
```
inspect?

tarsa commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -734,6 +734,26 @@ trait FlowOps[+Out, +Mat] {
+   * '''Cancels when''' downstream cancels
+   *
+   */
+  def inspect(f: Out ⇒ Unit): Repr[Out] = map(el ⇒ { f(el); el }).named("foreach")
```
named("inspect")?

tarsa commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInspectSpec.scala:
```diff
@@ -0,0 +1,95 @@
+    }
+  }
+
+  "A runForeach" must {
```
looks like the same tests are present in FlowForeachSpec.scala

tarsa reviewed (commented):
found some omissions

ktoso commented:
Thanks for review, addressed

ktoso commented:
Failure is unrelated and already has a ticket; https://jenkins.akka.io:8498/job/pr-validator-per-commit-jenkins/12002/testReport/junit/akka.stream.tck/EmptyProcessorTest/reproduce/

This is ready for review/merge I'd hope...

johanandren commented on akka-docs/src/main/paradox/stream/stages-overview.md:
```diff
@@ -719,6 +719,22 @@ Transform each element in the stream by calling a mapping function with it and p
+This operation is useful for performing inspecting the passed through element, usually by means of side-effecting operations (such as `println`, or emitting metrics), for each element without having to modify it.
+
+For logging signals (elements, completion, error) consider using the `log` stage instead,
+along with appropriate `ActorAttributes.createLogLevels`.
```
If we are going to have this, I'd like a big clear warning here, and in the API-docs that it cannot touch mutable state etc. because this is exactly how people are going to misuse it.

ktoso replied:
here you are: https://github.com/akka/akka/pull/24610/commits/577456f17c4cb8a4154f1cff4dd8e85574d641be

johanandren reviewed (commented):
I don't like it, but if we are going to have it, this LGTM with a big blinking warning sign attached as noted in comment.

ktoso commented:
Resolved conflict

ktoso commented:
Please merge @patriknw if this looks good by now.

patriknw commented on akka-docs/src/main/paradox/stream/stages-overview.md:
```diff
@@ -730,6 +730,31 @@ Transform each element in the stream by calling a mapping function with it and p
+
+Similar to `map`, however does not modify the passed through element, the returned value is ignored.
+
+This operation is useful for performing inspecting the passed through element, usually by means of side-effecting operations (such as `println`, or emitting metrics), for each element without having to modify it.
```
performing inspecting sounds funny, some other wording?, or simply s/performing inspecting/inspecting/

ktoso replied:
Reading these again it still "calls for" simply being called `foreach`.. and simply `performing side effects` wording here..

patriknw commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInspectSpec.scala:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    "complete future with failure when function throws" in assertAllStagesStopped {
+      val error = new Exception with NoStackTrace
```
there is a `TE` in streams testkit

patriknw commented on akka-stream/src/main/scala/akka/stream/javadsl/SubFlow.scala:
```diff
@@ -134,6 +134,27 @@ class SubFlow[In, Out, Mat](delegate: scaladsl.SubFlow[Out, Mat, scaladsl.Flow[I
+  /**
+   * Similar to [[map]], however does not modify the passed through element, the returned value is ignored.
+   *
+   * This operation is useful for performing inspecting the passed through element, usually by means of side-effecting
```
^

patriknw commented on akka-stream/src/main/scala/akka/stream/javadsl/Source.scala:
```diff
@@ -1072,6 +1072,27 @@ final class Source[Out, Mat](delegate: scaladsl.Source[Out, Mat]) extends Graph[
+  /**
+   * Similar to [[map]], however does not modify the passed through element, the returned value is ignored.
+   *
+   * This operation is useful for performing inspecting the passed through element, usually by means of side-effecting
```
^

patriknw commented on akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowInspectSpec.scala:
```diff
@@ -0,0 +1,57 @@
+  "A inspect" must {
+
+    "call the procedure for each element" in assertAllStagesStopped {
+      Source(1 to 3).inspect(testActor ! _).runWith(Sink.ignore) foreach {
```
wireTap isn't too bad:
```
Source(1 to 3).wireTap(Sink.foreach(testActor ! _)).runWith(Sink.ignore)
```

just saying that it's value is on the edge, but ok with me

patriknw reviewed (approved):
LGTM, some nitpicks

ktoso commented:
Thanks for review, will address soon

ktoso commented:
Addressed comments, though still https://github.com/akka/akka/pull/24610#discussion_r176302824 feeling that `foreach` inline naming would be best. If we don't think so, I'm ok with inspect I guess.

Rename to foreach, or merge at will.

patriknw commented:
Seems to be some strong opinions about the `foreach` name due to the non-terminal nature. I think `inspect` describes one typical use case well, but it doesn't necessarily cover all of them.

What do you think about adding an overload to `wireTap` that has signature:
```
def wireTap(f: Out ⇒ Unit): Repr[Out]
```

Then you can use that function for the simple wire tapping and the other wireTap for more advanced streaming wire tapping.

ktoso commented:
> def wireTap(f: Out ⇒ Unit): Repr[Out]

That is a great proposal, thanks! It also matches the semantics one may want/expect from such "side function" with regards to throwing exceptions, they should not kill the main stream, same as wireTap would not cancel the upstream. I will rename and impl using wireTap, and add the method to the new operators listing.

ktoso commented:
Error was my favourite whitesource tool ;) 

`WARNING: Invalid cookie header: "Set-Cookie: AWSALB=9Fx/EBEFxfXqZGdhjfarpJldO2BIvAmahokI/nBnMlK1h3xQUH6z5PjkzLliqn5IidD4XygoROZfJOXU30uKX9OvJKEaHVu8rYq6yey4D0bvaFk5x9PMk5+lSXPm; Expires=Mon, 30 Apr 2018 16:30:50 GMT; Path=/". Invalid 'expires' attribute: Mon, 30 Apr 2018 16:30:50 GMT`

ktoso commented:
Was debated and reviewed to death, the new name makes a ton of sense wireTap it is then!

Also automagically worked with existing docs gen, since wireTaps exist, so it only added the new signature there -- another great upside

patriknw commented on akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala:
```diff
@@ -483,6 +483,29 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
     new Flow(delegate.map(f.apply))
 
+  /**
+   * Similar to [[map]], however does not modify the passed through element, the returned value is ignored.
```
I also find it strange to refer to `map` since the purpose is completely different

patriknw commented on akka-stream/src/main/scala/akka/stream/javadsl/SubFlow.scala:
```diff
@@ -135,6 +135,29 @@ class SubFlow[In, Out, Mat](delegate: scaladsl.SubFlow[Out, Mat, scaladsl.Flow[I
 
+  /**
+   * Similar to [[map]], however does not modify the passed through element, the returned value is ignored.
+   * This is a simplified version of `wireTap(Sink)`, which you may use to wireTap a Sink onto this stream.
```
"wireTap a Sink"
copy pasta?

ktoso replied:
was intentional wording actually, "to wire tap" used as a verb... maybe not so natural?

> What does it mean to wire tap?
> an act or instance of tapping telephone or telegraph wires for evidence or other information. ... to listen in on by means of a wiretap: to wiretap a telephone; to wiretap a conversation.

patriknw reviewed (commented):
I like the outcome of this, some post merging nitpicks
